src/revlanguage/functions/phylogenetics/ratematrix/Func_InvModel.cpp

```diff
@@ -0,0 +1,153 @@
+#include "Func_InvModel.h"
+
+#include "RealPos.h"
+#include "RlDeterministicNode.h"
+#include "RlMixtureModel.h"
+#include "RlRateMatrix.h"
+#include "RlSimplex.h"
+#include "GenericFunction.h"
+#include "TypedDagNode.h"
+#include "Argument.h"
+#include "ArgumentRule.h"
+#include "ArgumentRules.h"
+#include "RbException.h"
+#include "RevVariable.h"
+#include "RlFunction.h"
+#include "RlBoolean.h"
+#include "Integer.h"
+#include "Simplex.h"
+#include "TypeSpec.h"
+
+#include "DistributionChisq.h"
+#include "RbMathFunctions.h"
+#include "ConcreteTimeReversibleRateMatrix.h"
+#include "ConcreteMixtureModel.h"
+#include "UnitMixtureModel.h"
+
+using std::vector;
+using std::unique_ptr;
+
+// wait, so, how do we encode the operations like scale(model), rate(model), etc?
+// well, ratemodel should perhaps be its OWN interface!
+
+// how about things like the m3 model?
+// also, could we create this using a FUNCTION?
+
+RevBayesCore::ConcreteMixtureModel* InvModelFunc(const RevBayesCore::SiteMixtureModel& sub_model, double p)
+{
+    using namespace RevBayesCore;
+
+    if (p < 0 or p > 1)
+        throw RbException()<<"fnInv: pInv should be in [0,1], but pInv = "<<p;
+
+    vector<unique_ptr<RevBayesCore::SiteMixtureModel>> sub_models;
+    vector<double> fractions = {1-p, p};
+
+    // The non-invariant part.
+    sub_models.push_back( unique_ptr<RevBayesCore::SiteMixtureModel>(sub_model.clone()) );
+
+    auto f = sub_model.componentProbs();
+    int n = sub_model.getNumberOfStates();
+    vector<double> pi(n,0);
+    for(int m=0; m<f.size(); m++)
+    {
+        auto sub_pi = sub_model.getRootFrequencies(m);
+        for(int l=0;l<pi.size();l++)
+            pi[l] += sub_pi[l] * f[m];
+    }
+    vector<double> er(n*(n-1)/2,0);
+    RevBayesCore::ConcreteTimeReversibleRateMatrix INV(er,pi,{});
+
+    sub_models.push_back( unique_ptr<RevBayesCore::SiteMixtureModel>(new RevBayesCore::UnitMixtureModel(INV)) );
+    
+    auto inv_model = new RevBayesCore::ConcreteMixtureModel(sub_models, fractions);
+    inv_model->setRate(1);
```

Comment:
I've changed this so that instead of setting the rate to 1, we set the rate equal to the submodel rate.  This is equivalent to what revbayes is doing now.